TUZ-145 Use ONNX node id in span in frontend

areusch opened the pull request:
* Logs an identifying name to an IRModule attribute "tvm-frontend-op-ids"
 * When possible (it's legal for an ONNX NodeProto to have an empty
   name), uses the name from the NodeProto to identify the op. When
   name is empty, uses the index of the NodeProto.

YuchenJin reviewed (commented):
Thank you @junrushao1994 for improving the code quality! I add the preference of methods to use to register global function to the [checklist](https://github.com/tlc-pack/relax/wiki/Relax-Development-Guidelines#checklist) of the dev guideline.

The CI and lint will be set up thanks to @areusch and @konturn (I need to modify the jenkinsfile). I believe we will have more formatting stuff to do to pass the lint checks. :)

junrushao1994 commented:
> C6. Global packed function registration: Preference: set_body_method<> > set_body_typed > set_body. Only use set_body when the function accepts variable-length arguments because it's hard to reason about and document, as well as requiring extra code in both Python and C++.

Not necessarily `set_body_method > set_body_typed > set_body` though. I think we could prefer whichever is shorter and more succinct :-)

junrushao1994 commented:
BTW, the parser fails on synr 0.6 with:

```
AttributeError: 'TypeApply' object has no attribute 'id'
```

Having synr as a separate repo does introduce these synchronization issues :-(

junrushao1994 commented on src/relax/vm/builtin.cc:
```diff
@@ -70,58 +65,54 @@ TVM_REGISTER_GLOBAL("vm.builtin.load_shape")
+      return storage;
+    });
+
+TVM_REGISTER_GLOBAL("vm.builtin.alloc_tensor").set_body_method<Storage>(&StorageObj::AllocNDArray);
```
Ahh I see. That sounds tricky though :-(

Is there any particular reason the order differs? If not, do you think we should somehow unify them to avoid future confusion?

junrushao1994 replied:
Just noticed that shape-first seems to be our convention in VM codegen. I will change the parameters of `StorageObj::AllocNDArray` instead

YuchenJin replied:
There is no particular reason and I think we should unify them by modifying `EmitAllocTensor` in CodeGenVM. :)

junrushao1994 replied:
Ok will do :-)

YuchenJin replied:
> Just noticed that shape-first seems to be our convention in VM codegen. I will change the parameters of `StorageObj::AllocNDArray` instead

Oh just saw this comment, I think either changing `StorageObj::AllocNDArray` or `EmitAllocTensor` is good. :) Some unit tests (e.g. [test_vm_constant_serialize](https://github.com/tlc-pack/relax/blob/5527a69dc50dc051b819044d4575f01e0a301b0f/tests/python/relax/test_vm.py#L136) ) which directly calls `vm.builtin.alloc_tensor` need to be updated if the AllocNDArray args order is changed.

YuchenJin commented:
> Not necessarily `set_body_method > set_body_typed > set_body` though. I think we could prefer whichever is shorter and more succinct :-)

Sounds great! Let me update it. :)

YuchenJin commented:
> BTW, the parser fails on synr 0.6 with:
> 
> ```
> AttributeError: 'TypeApply' object has no attribute 'id'
> ```
> 
> Having synr as a separate repo does introduce these synchronization issues :-(

Yeah, currently the parser has this error with synr 0.6 but works fine with synr 0.5. I didn't dive into the cause yet.

junrushao1994 commented:
@YuchenJin fixed according to your comments: https://github.com/tlc-pack/relax/pull/65#discussion_r785484553

junrushao1994 commented:
All unittest should pass now with synr===0.6

YuchenJin reviewed (approved):
LGTM! Thanks @junrushao1994!

areusch committed:
Make attached spans uniquely identify ONNX operators with names.

 * Logs an identifying name to an IRModule attribute "tvm-frontend-op-ids"
 * When possible (it's legal for an ONNX NodeProto to have an empty
   name), uses the name from the NodeProto to identify the op. When
   name is empty, uses the index of the NodeProto.

areusch commented:
@jwfromm @driazati can you have a look? I've changed this PR to use the NodeProto name field when it's not empty (it [can be](https://onnx.ai/onnx/api/helper.html#onnx.helper.make_node)) and the index otherwise. In practice, the node name should be enough.

driazati commented on python/tvm/relax/frontend/onnx/onnx_frontend.py:
```diff
@@ -2032,8 +2059,9 @@ def _convert_operator(
             Operator name, such as Convolution, FullyConnected
-        node_index : int
-            Index of the node in the ONNX graph.
+        node_id : [str, int]
```
```suggestion
        node_id : Union[str, int]
```
